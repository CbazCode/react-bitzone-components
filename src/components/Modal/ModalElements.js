import styled, { keyframes } from 'styled-components';
import { MdClose } from 'react-icons/md';

const entry = keyframes`
  from{
      /* transform: scale(0.2); */
      opacity:0;
      background: rgba(0, 0, 0, 0);
  }
  to{
      /* transform: scale(1); */
      opacity:1;
      background: rgba(0, 0, 0, 0.8);
      
      
  }
`;

export const Background = styled.div`
  width: 100vw;
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: flex;
  animation: ${entry} 0.6s ease-in-out forwards;
  justify-content: center;
  align-items: center;
  overflow-y: auto;
  z-index: 999;
`;

const widthModalWrapper = (withImage) => {
  return withImage ? '850px' : '400px';
};

const gridModalWrapper = (withImage) => {
  return withImage ? '1fr 1fr' : '1fr';
};

export const ModalWrapper = styled.div`
  width: ${({ withImage }) => widthModalWrapper(withImage)};
  font-family: 'Sora', sans-serif;
  box-shadow: 0 5px 16px rgba(0, 0, 0, 0.2);
  background: #fff;
  color: #000;
  display: grid;
  /* margin:auto 2rem; */
  grid-template-columns: ${({ withImage }) => gridModalWrapper(withImage)};
  position: relative;
  z-index: 10;
  border-radius: 10px;

  @media screen and (max-width: 930px) {
    grid-template-columns: 1fr;
    width: 400px;
  }
`;

export const ModalImg = styled.img`
  width: 100%;
  height: 100%;
  border-radius: 0 10px 10px 0px;
  background: #000;
  @media screen and (max-width: 930px) {
    border-radius: 0 0 10px 10px;
  }
`;

export const ModalContent = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  line-height: 1.8;
  color: #141414;
  padding: 1rem;
`;

export const ModalBody = styled.p`
  padding: 2rem;
`;

export const ModalHeader = styled.h1`
  font-size: 2rem;
`;

export const CloseModalButton = styled(MdClose)`
  cursor: pointer;
  position: absolute;
  top: 20px;
  right: 20px;
  width: 32px;
  height: 32px;
  padding: 0;
  z-index: 10;
`;
