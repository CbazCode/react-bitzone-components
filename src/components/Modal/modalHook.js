import { useState } from "react";

const useModal = () => {
    const [showModal, setShowModal] = useState(false);

    const openModal = () => {
      setShowModal(prev => !prev);
    };

    return{showModal, openModal, setShowModal}
}

export default useModal;