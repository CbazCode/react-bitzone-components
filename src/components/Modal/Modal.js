import React, { useRef, useEffect, useCallback } from 'react';
import { useSpring, animated } from 'react-spring';
import { Background, CloseModalButton,  ModalContent, ModalImg, ModalWrapper } from './ModalElements';



const Modal = ({ showModal, setShowModal, children,appear, withImage, sourceImage}) => { 
    const modalRef = useRef(); 

    const selectTransform = (appear, showModal) => {
        switch (appear) {
            case "top":
                return showModal ? `translateY(0%)` : `translateY(-100%)`
            case "bottom":
                return showModal ? `translateY(0%)` : `translateY(100%)`
            case "right":
                return showModal ? `translateX(0%)` : `translateX(100%)`
            case "left":
                return showModal ? `translateX(0%)` : `translateX(-100%)`    
            case "fadeIn":
                return showModal ? `none` : `none`         
            default:
                return showModal ? `translateY(0%)` : `translateY(-100%)`
        }
    }
  
    const animation = useSpring({
      config: {
        duration: 500
      },
      opacity: showModal ? 1 : 0,
      // transform: showModal ? `scale(1)` : `scale(0.5)`
      transform: selectTransform(appear, showModal),
    });
  
    const closeModal = e => {
      if (modalRef.current === e.target) {
        setShowModal(false);
      }
    };
  
    const keyPress = useCallback(
      e => {
        if (e.key === 'Escape' && showModal) {
          setShowModal(false);
          console.log('I pressed');
        }
      },
      [setShowModal, showModal]
    );
  
    useEffect(
      () => {
        document.addEventListener('keydown', keyPress);
        return () => document.removeEventListener('keydown', keyPress);
      },
      [keyPress]
    );
  
    return (
      <>
        {showModal ? (
          <Background onClick={closeModal} ref={modalRef}>
            <animated.div style={animation}>
              <ModalWrapper  withImage={withImage}>
                  
                <ModalContent>
                  {children}
                </ModalContent>
                {
                  withImage &&
                  <ModalImg src={sourceImage} alt='camera' />// eslint-disable-line no-unused-vars
                }
                <CloseModalButton
                  aria-label='Close modal'
                  onClick={() => setShowModal(prev => !prev)}
                />
              </ModalWrapper>
            </animated.div>
          </Background>
        ) : null}
      </>
    );
  };

  export default Modal;