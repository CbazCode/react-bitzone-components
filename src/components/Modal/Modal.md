```js
import useModal from './modalHook.js'
import {ModalBody, ModalHeader} from './ModalElements.js'
import Modal from './Modal'
import Button from '../Button/Button.js'

const {showModal, openModal, setShowModal} = useModal();


<>
    <Button 
        bgColor="danger"  size="md" rounded="soft" color="white" onClick={openModal}>
        I'am modal
    </Button>
    
    <Modal showModal={showModal} setShowModal={setShowModal}  >
            <ModalHeader>Are you ready?</ModalHeader>
            <ModalBody>Get exclusive access to our next launch. Lorem ipsum dolor sit amet, consectetur adipisicing elit. Veritatis minima ratione placeat facere illo fuga non id aliquam cumque molestias! </ModalBody>
            <Button bgColor="info" size="xl" rounded="full" color="white">
                Button
            </Button>
    </Modal>
</>
```

```js
import useModal from './modalHook.js'
import {ModalBody, ModalHeader} from './ModalElements.js'
import Modal from './Modal'
import Button from '../Button/Button.js'
import img from './modal.jpg'

const {showModal, openModal, setShowModal} = useModal();


<>
    <Button 
        bgColor="primary"  size="md" rounded="soft" color="white" onClick={openModal}>
        I'am modal
    </Button>
    
    <Modal showModal={showModal} setShowModal={setShowModal}  withImage={true} sourceImage={img}>
            <ModalHeader>Are you ready?</ModalHeader>
            <ModalBody>Get exclusive access to our next launch. Lorem ipsum dolor sit amet, consectetur adipisicing elit. Veritatis minima ratione placeat facere illo fuga non id aliquam cumque molestias! </ModalBody>
            <Button bgColor="info" size="xl" rounded="full" color="white">
                Button
            </Button>
    </Modal>
</>
```

```js
import useModal from './modalHook.js'
import {ModalBody, ModalHeader} from './ModalElements.js'
import Modal from './Modal'
import Button from '../Button/Button.js'

const {showModal, openModal, setShowModal} = useModal();


<>
    <Button 
        bgColor="secondary"  size="md" rounded="soft" color="white" onClick={openModal}>
        I'am modal
    </Button>
    
    <Modal showModal={showModal} setShowModal={setShowModal}  appear="right">
            <ModalHeader>Are you ready?</ModalHeader>
            <ModalBody>Get exclusive access to our next launch. Lorem ipsum dolor sit amet, consectetur adipisicing elit. Veritatis minima ratione placeat facere illo fuga non id aliquam cumque molestias! </ModalBody>
            <Button bgColor="info" size="xl" rounded="full" color="white">
                Button
            </Button>
    </Modal>
</>
```