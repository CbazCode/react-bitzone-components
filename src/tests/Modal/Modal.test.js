
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import Modal from '../../components/Modal/Modal'

describe('Pruebas en el componente <Modal/>', () => {

    test('debe renderizarse el componente correctamente', () => {
        const { container } = render(<Modal showModal={true}/>);
        expect(container.firstChild).toMatchSnapshot();
    })
})
