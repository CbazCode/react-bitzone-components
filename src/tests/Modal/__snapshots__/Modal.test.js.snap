// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <Modal/> debe renderizarse el componente correctamente 1`] = `
.c0 {
  width: 100vw;
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-animation: iwwYms 0.6s ease-in-out forwards;
  animation: iwwYms 0.6s ease-in-out forwards;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow-y: auto;
  z-index: 999;
}

.c1 {
  width: 400px;
  font-family: 'Sora',sans-serif;
  box-shadow: 0 5px 16px rgba(0,0,0,0.2);
  background: #fff;
  color: #000;
  display: grid;
  grid-template-columns: 1fr;
  position: relative;
  z-index: 10;
  border-radius: 10px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 1.8;
  color: #141414;
  padding: 1rem;
}

.c3 {
  cursor: pointer;
  position: absolute;
  top: 20px;
  right: 20px;
  width: 32px;
  height: 32px;
  padding: 0;
  z-index: 10;
}

@media screen and (max-width:930px) {
  .c1 {
    grid-template-columns: 1fr;
    width: 400px;
  }
}

<div
  class="c0"
>
  <div
    style="opacity: 1; transform: translateY(0%);"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      />
      <svg
        aria-label="Close modal"
        class="c3"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
  </div>
</div>
`;
